ocean/helium/snapshots.py

```diff
@@ -138,3 +143,263 @@ def download_bytes_from_snapshot(src, snapshothash):
+            entry = {
+                'hash': hash_file(f),
+                'size': os.path.getsize(f),
+                'user_meta': {},
+                'physical_keys': [{
+                    'type': 'local_file',
+                    'path': os.path.abspath(f)
+                }]
+            }
+            snap._data[str(f)] = entry
+        return snap
+
+    def get(self, logical_key):
+        """
+        Gets object from local_key and returns it as an in-memory object.
+
+        Args:
+            logical_key: logical key of the object to get
+
+        Returns:
+            A deserialized object from the logical_key or
+                a stream of bytes if deserialization info is missing
+
+        Raises:
+            KeyError: when logical_key is not present in the snapshot
+            physical key failure
+            hash verification fail
+        """
+        entry = self._data[logical_key]
+        physical_keys = entry['physical_keys']
+        physical_key = physical_keys[0] # TODO: support multiple physical keys
+        stream = dereference_physical_key(physical_key)
+        # TODO: verify hash
+        if 'target' in entry:
+            # TODO: dispatch on target to deserialize
+            raise NotImplementedError
+
+        return stream
+
+    def get_file(self, logical_key, path):
+        """
+        Gets object from logical_key inside the snapshot and saves it to path.
+
+        Args:
+            logical_key: logical key inside snapshot to get
+            path: where to put the file
+
+        Returns:
+            None
+
+        Raises:
+            logical key not found
+            physical key failure
+            fail to create file
+            fail to finish write
+        """
+        raise NotImplementedError
+
+    def get_meta(self, logical_key):
+        """
+        Returns user metadata for specified logical key.
+        """
+        entry = self._data[logical_key]
+        return entry['user_meta']
+
+    def dump(self, path):
+        """
+        Serializes this snapshot to a file at path.
+
+        Args:
+            path: where to serialize the snapshot to
+
+        Returns:
+            None
+
+        Raises:
+            fail to create file
+            fail to finish write
+        """
+        snap = Snapshot()
+        with jsonlines.open(path, mode='w') as writer:
+            writer.write(snap._meta)
+            for logical_key, obj in self._data.items():
+                obj['logical_key'] = logical_key
+                writer.write(obj)
+
+    def update(self, logical_key, entry):
+        """
+        Returns a new snapshot with the object at logical_key set to entry.
+
+        Args:
+            logical_key: logical key to update
+            entry: new entry to place at logical_key in the snapshot
+
+        Returns:
+            A new snapshot
+        """
+        snap = self._clone()
+        # TODO validate entry contents
+        snap._data[logical_key] = entry
```

Comment:
If the value is None then remove it

Reply:
that's delete -- if you want we can merge that into one API call, but right now delete is the clear way to remove logical_keys from a package